**Pandas -> Data Manipulation**



In [ ]:
import numpy as np
import pandas as pd

In [ ]:
len(dir(pd))

141

**In Pandas Library we have 141 Built_in_functions from 141 we are going to learn the Functions which are used in AI**

- `Series `
- `DataFrame`
- `shape `
- `size `
- `loc `
- `iloc `
- `rename `
- `fillna `
- `dropna `
- `columns `
- `index `
- `drop `
- `axis = 1`
- `axis = 0`
- `isnull().sum()`
- `unique `
- `value_counts()`
- `select_dtypes `
- `info `
- `head `
- `tail `
- `sample `
- `groupBy`
- `concat `


In [ ]:
# Series built_in_function in pandas will create only 1 column (1D data)

a = np.array([10,12,14,20,22]) # 1D

b = pd.Series(a , index = ['r1','r2','r3','r4','r5']) # giving 1D data to Series -> so Series will create one column represtation

b

,0
r1,10
r2,12
r3,14
r4,20
r5,22


In [ ]:
b['r3']

np.int64(14)

In [ ]:
# DataFrame -> Proper Excel format -> we can create (n) number of rows and columns

a = {'Name' : ['John' , 'peter' , 'steve' , 'brad' , 'smith' , 'Hussy'] ,
     'Age' : [27 , 30,32,35,29,41] ,
     'Experience' : [1,2,3,4,5,6] ,
     'Company' : ['Google' , 'Meta' , 'IBM' , 'MSFT' , 'Wipro' , 'Cognizant'],
     'Salary' : [1200 , 1400 , 1000 , 1800 , 1100 , 900]
     }

b = pd.DataFrame(a , index = ['r1','r2','r3','r4','r5' , 'r6']) # give the dict to dataframe so excel will create
b

,Name,Age,Experience,Company,Salary
r1,John,27,1,Google,1200
r2,peter,30,2,Meta,1400
r3,steve,32,3,IBM,1000
r4,brad,35,4,MSFT,1800
r5,smith,29,5,Wipro,1100
r6,Hussy,41,6,Cognizant,900


In [ ]:
b.loc['r3' : 'r5' , 'Name' : 'Company']

,Name,Age,Experience,Company
r3,steve,32,3,IBM
r4,brad,35,4,MSFT
r5,smith,29,5,Wipro


In [ ]:
b.iloc[2 : 5 , 0 : 4]

,Name,Age,Experience,Company
r3,steve,32,3,IBM
r4,brad,35,4,MSFT
r5,smith,29,5,Wipro


In [ ]:
b.iloc[1 : 5 , 1 : 4]

,Age,Experience,Company
r2,30,2,Meta
r3,32,3,IBM
r4,35,4,MSFT
r5,29,5,Wipro


In [ ]:
b.loc['r2' : 'r5' , 'Age' : 'Company']

,Age,Experience,Company
r2,30,2,Meta
r3,32,3,IBM
r4,35,4,MSFT
r5,29,5,Wipro


In [ ]:
b.loc['r2' : 'r4' , 'Experience' : 'Salary']

,Experience,Company,Salary
r2,2,Meta,1400
r3,3,IBM,1000
r4,4,MSFT,1800


In [ ]:
b.loc['r2' : 'r5' , 'Age' : 'Company']

,Age,Experience,Company
r2,30,2,Meta
r3,32,3,IBM
r4,35,4,MSFT
r5,29,5,Wipro


In [ ]:
b['Company']['r2' : 'r4']

,Company
r2,Meta
r3,IBM
r4,MSFT


In [ ]:
b['Age']['r4']

np.int64(35)

In [ ]:
b['Company']['r3']

'IBM'

In [ ]:
b

,Name,Age,Experience,Company,Salary
r1,John,27,1,Google,1200
r2,peter,30,2,Meta,1400
r3,steve,32,3,IBM,1000
r4,brad,35,4,MSFT,1800
r5,smith,29,5,Wipro,1100
r6,Hussy,41,6,Cognizant,900


In [ ]:
b.loc[ : , 'Company' : 'Salary']

,Company,Salary
r1,Google,1200
r2,Meta,1400
r3,IBM,1000
r4,MSFT,1800
r5,Wipro,1100
r6,Cognizant,900


In [ ]:
b

,Name,Age,Experience,Company,Salary
r1,John,27,1,Google,1200
r2,peter,30,2,Meta,1400
r3,steve,32,3,IBM,1000
r4,brad,35,4,MSFT,1800
r5,smith,29,5,Wipro,1100
r6,Hussy,41,6,Cognizant,900


In [ ]:
b.iloc[ : , 3 : ]

,Company,Salary
r1,Google,1200
r2,Meta,1400
r3,IBM,1000
r4,MSFT,1800
r5,Wipro,1100
r6,Cognizant,900


In [ ]:
# loc and iloc syntax = [RS : RE , CS : CE]
# loc will not follow index and n-1 logic
# iloc will follow index and n-c logic

In [ ]:
b

,Name,Age,Experience,Company,Salary
r1,John,27,1,Google,1200
r2,peter,30,2,Meta,1400
r3,steve,32,3,IBM,1000
r4,brad,35,4,MSFT,1800
r5,smith,29,5,Wipro,1100
r6,Hussy,41,6,Cognizant,900


In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
b['Company']['r3'] = 'VIHARA TECH'

In [ ]:
b

,Name,Age,Experience,Company,Salary
r1,John,27,1,Google,1200
r2,peter,30,2,Meta,1400
r3,steve,32,3,VIHARA TECH,1000
r4,brad,35,4,MSFT,1800
r5,smith,29,5,Wipro,1100
r6,Hussy,41,6,Cognizant,900


In [ ]:
b['Age'] = [11,12,13,14,15,16]

In [ ]:
b

,Name,Age,Experience,Company,Salary
r1,John,11,1,Google,1200
r2,peter,12,2,Meta,1400
r3,steve,13,3,VIHARA TECH,1000
r4,brad,14,4,MSFT,1800
r5,smith,15,5,Wipro,1100
r6,Hussy,16,6,Cognizant,900


In [ ]:
b['Tax'] = [12,14,9,18,6,22]


In [ ]:
b

,Name,Age,Experience,Company,Salary,Tax
r1,John,11,1,Google,1200,12
r2,peter,12,2,Meta,1400,14
r3,steve,13,3,VIHARA TECH,1000,9
r4,brad,14,4,MSFT,1800,18
r5,smith,15,5,Wipro,1100,6
r6,Hussy,16,6,Cognizant,900,22


In [ ]:
# NumPy
# axis = 1 -> row
# axis = 0 -> column

# Pandas
# axis = 1 -> column
# axis = 0 -> row

In [ ]:
f = 'india'

f = f.capitalize()

print(f)

India


In [ ]:
b

,Name,Age,Experience,Company,Salary,Tax
r1,John,11,1,Google,1200,12
r2,peter,12,2,Meta,1400,14
r3,steve,13,3,VIHARA TECH,1000,9
r4,brad,14,4,MSFT,1800,18
r5,smith,15,5,Wipro,1100,6
r6,Hussy,16,6,Cognizant,900,22


In [ ]:
b = b.drop(['Company']  , axis= 1)

In [ ]:
b

,Name,Age,Experience,Salary,Tax
r1,John,11,1,1200,12
r2,peter,12,2,1400,14
r3,steve,13,3,1000,9
r4,brad,14,4,1800,18
r5,smith,15,5,1100,6
r6,Hussy,16,6,900,22


In [ ]:
b = b.drop(['r3'] , axis = 0)

In [ ]:
b

,Name,Age,Experience,Salary,Tax
r1,John,11,1,1200,12
r2,peter,12,2,1400,14
r4,brad,14,4,1800,18
r5,smith,15,5,1100,6
r6,Hussy,16,6,900,22


In [ ]:
b = b.rename(columns= {'Tax' : 'EMI'})

In [ ]:
b

,Name,Age,Experience,Salary,EMI
r1,John,11,1,1200,12
r2,peter,12,2,1400,14
r4,brad,14,4,1800,18
r5,smith,15,5,1100,6
r6,Hussy,16,6,900,22


In [ ]:
b['TDS'] = [200 , np.nan , 300 , np.nan , 400]


In [ ]:
b

,Name,Age,Experience,Salary,EMI,TDS
r1,John,11,1,1200,12,200.0
r2,peter,12,2,1400,14,NaN
r4,brad,14,4,1800,18,300.0
r5,smith,15,5,1100,6,NaN
r6,Hussy,16,6,900,22,400.0


In [ ]:
b['TDS'] = b['TDS'].fillna(100)

In [ ]:
b.to_csv("Sample_Data_AI_class.csv",index = False)

In [ ]:
b

,Name,Age,Experience,Salary,EMI,TDS
r1,John,11,1,1200,12,200.0
r2,peter,12,2,1400,14,100.0
r4,brad,14,4,1800,18,300.0
r5,smith,15,5,1100,6,100.0
r6,Hussy,16,6,900,22,400.0


In [ ]:
b

,Name,Age,Experience,Salary,EMI,TDS
r1,John,11,1,1200,12,200.0
r2,peter,12,2,1400,14,100.0
r4,brad,14,4,1800,18,300.0
r5,smith,15,5,1100,6,100.0
r6,Hussy,16,6,900,22,400.0


In [ ]:
b['Balance_Amount'] = [10,np.nan,20,np.nan,30]

In [ ]:
b

,Name,Age,Experience,Salary,EMI,TDS,Balance_Amount
r1,John,11,1,1200,12,200.0,10.0
r2,peter,12,2,1400,14,100.0,NaN
r4,brad,14,4,1800,18,300.0,20.0
r5,smith,15,5,1100,6,100.0,NaN
r6,Hussy,16,6,900,22,400.0,30.0


In [ ]:
b = b.dropna()

In [ ]:
b

,Name,Age,Experience,Salary,EMI,TDS,Balance_Amount
r1,John,11,1,1200,12,200.0,10.0
r4,brad,14,4,1800,18,300.0,20.0
r6,Hussy,16,6,900,22,400.0,30.0


`In the last class we created our own data and implemented many operations -> know I am going to load the data from the excel sheet and then I will do the all operations `

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('/content/Titanic-Dataset (1).csv')

In [4]:
df.shape

(891, 12)

In [11]:
df.head(2) # will show first 5

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C


In [12]:
df.tail() # last 5 rows

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [15]:
df.sample(2) # out of 891 randomlly 5 will be given

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
372,373,0,3,"Beavan, Mr. William Thomas",male,19.0,0,0,323951,8.0500,NaN,S
650,651,0,3,"Mitkoff, Mr. Mito",male,NaN,0,0,349221,7.8958,NaN,S


In [16]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [17]:
df.index

RangeIndex(start=0, stop=891, step=1)

In [19]:
df.info()  # complete information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [20]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


In [21]:
df = df.dropna()

In [22]:
df.shape

(183, 12)

In [23]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [24]:
df = df.drop(['PassengerId' , 'Name' , 'Ticket'] , axis = 1)

In [25]:
df

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin,Embarked
1,1,1,female,38.0,1,0,71.2833,C85,C
3,1,1,female,35.0,1,0,53.1000,C123,S
6,0,1,male,54.0,0,0,51.8625,E46,S
10,1,3,female,4.0,1,1,16.7000,G6,S
11,1,1,female,58.0,0,0,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...
871,1,1,female,47.0,1,1,52.5542,D35,S
872,0,1,male,33.0,0,0,5.0000,B51 B53 B55,S
879,1,1,female,56.0,0,1,83.1583,C50,C
887,1,1,female,19.0,0,0,30.0000,B42,S


In [26]:
df.shape

(183, 9)

In [28]:
df.size

1647

In [27]:
183 * 9

1647

In [29]:
df = df.rename(columns={'Fare' : 'Price'})

In [30]:
df

,Survived,Pclass,Sex,Age,SibSp,Parch,Price,Cabin,Embarked
1,1,1,female,38.0,1,0,71.2833,C85,C
3,1,1,female,35.0,1,0,53.1000,C123,S
6,0,1,male,54.0,0,0,51.8625,E46,S
10,1,3,female,4.0,1,1,16.7000,G6,S
11,1,1,female,58.0,0,0,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...
871,1,1,female,47.0,1,1,52.5542,D35,S
872,0,1,male,33.0,0,0,5.0000,B51 B53 B55,S
879,1,1,female,56.0,0,1,83.1583,C50,C
887,1,1,female,19.0,0,0,30.0000,B42,S


In [31]:
df.shape

(183, 9)

In [32]:
df['Sex'].unique()

array(['female', 'male'], dtype=object)

In [34]:
df['Sex'].value_counts()

,count
Sex,
male,95
female,88


In [35]:
95 + 88

183

In [33]:
df['Embarked'].unique()

array(['C', 'S', 'Q'], dtype=object)

In [36]:
df['Embarked'].value_counts()

,count
Embarked,
S,116
C,65
Q,2


In [38]:
df

,Survived,Pclass,Sex,Age,SibSp,Parch,Price,Cabin,Embarked
1,1,1,female,38.0,1,0,71.2833,C85,C
3,1,1,female,35.0,1,0,53.1000,C123,S
6,0,1,male,54.0,0,0,51.8625,E46,S
10,1,3,female,4.0,1,1,16.7000,G6,S
11,1,1,female,58.0,0,0,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...
871,1,1,female,47.0,1,1,52.5542,D35,S
872,0,1,male,33.0,0,0,5.0000,B51 B53 B55,S
879,1,1,female,56.0,0,1,83.1583,C50,C
887,1,1,female,19.0,0,0,30.0000,B42,S


In [39]:
t = pd.DataFrame({'Animal' : ['Lion','tiger','Lion','tiger','Lion','tiger'] ,
                  'Speed' : [100 , 180 , 90 , 190 , 80 , 200]
                  }
                 )

In [40]:
t

,Animal,Speed
0,Lion,100
1,tiger,180
2,Lion,90
3,tiger,190
4,Lion,80
5,tiger,200


In [41]:
t.groupby(['Animal']).mean()

,Speed
Animal,
Lion,90.0
tiger,190.0


- `I loaded titanic data into df varibale -> when I loaded in df we have 891 rows and 12 columns `

- `but afer cleaning i have 183 rows and 9 columns this 183 rows and 9 columns present in df varibale `

- `we are doing the changes in df only it is not going to do the change in the original data in the original dataset it will be 891  rows and 12 columns only `

- `if we want we need to save the data which is present in df into other excel sheet `



In [43]:
df.to_csv("Titanic_cleaned_data.csv",index=False)

In [44]:
df

,Survived,Pclass,Sex,Age,SibSp,Parch,Price,Cabin,Embarked
1,1,1,female,38.0,1,0,71.2833,C85,C
3,1,1,female,35.0,1,0,53.1000,C123,S
6,0,1,male,54.0,0,0,51.8625,E46,S
10,1,3,female,4.0,1,1,16.7000,G6,S
11,1,1,female,58.0,0,0,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...
871,1,1,female,47.0,1,1,52.5542,D35,S
872,0,1,male,33.0,0,0,5.0000,B51 B53 B55,S
879,1,1,female,56.0,0,1,83.1583,C50,C
887,1,1,female,19.0,0,0,30.0000,B42,S


In [45]:
num_cols = df.select_dtypes(exclude = 'object')


In [46]:
num_cols

,Survived,Pclass,Age,SibSp,Parch,Price
1,1,1,38.0,1,0,71.2833
3,1,1,35.0,1,0,53.1000
6,0,1,54.0,0,0,51.8625
10,1,3,4.0,1,1,16.7000
11,1,1,58.0,0,0,26.5500
...,...,...,...,...,...,...
871,1,1,47.0,1,1,52.5542
872,0,1,33.0,0,0,5.0000
879,1,1,56.0,0,1,83.1583
887,1,1,19.0,0,0,30.0000


In [47]:
text_cols = df.select_dtypes(include='object')

In [48]:
text_cols

,Sex,Cabin,Embarked
1,female,C85,C
3,female,C123,S
6,male,E46,S
10,female,G6,S
11,female,C103,S
...,...,...,...
871,female,D35,S
872,male,B51 B53 B55,S
879,female,C50,C
887,female,B42,S


In [49]:
# concat

t = pd.DataFrame({'Animal' : ['Lion','tiger','Lion','tiger','Lion','tiger'] ,
                  'Speed' : [100 , 180 , 90 , 190 , 80 , 200]
                  }
                 )
t

,Animal,Speed
0,Lion,100
1,tiger,180
2,Lion,90
3,tiger,190
4,Lion,80
5,tiger,200


In [52]:
a = {'Name' : ['John' , 'peter' , 'steve' , 'brad' , 'smith' , 'Hussy'] ,
     'Age' : [27 , 30,32,35,29,41] ,
     'Experience' : [1,2,3,4,5,6] ,
     'Company' : ['Google' , 'Meta' , 'IBM' , 'MSFT' , 'Wipro' , 'Cognizant'],
     'Salary' : [1200 , 1400 , 1000 , 1800 , 1100 , 900]
     }

b = pd.DataFrame(a)
b

,Name,Age,Experience,Company,Salary
0,John,27,1,Google,1200
1,peter,30,2,Meta,1400
2,steve,32,3,IBM,1000
3,brad,35,4,MSFT,1800
4,smith,29,5,Wipro,1100
5,Hussy,41,6,Cognizant,900


In [53]:
t

,Animal,Speed
0,Lion,100
1,tiger,180
2,Lion,90
3,tiger,190
4,Lion,80
5,tiger,200


In [54]:
b

,Name,Age,Experience,Company,Salary
0,John,27,1,Google,1200
1,peter,30,2,Meta,1400
2,steve,32,3,IBM,1000
3,brad,35,4,MSFT,1800
4,smith,29,5,Wipro,1100
5,Hussy,41,6,Cognizant,900


In [55]:
final_data = pd.concat([t , b] , axis=1)

In [56]:
final_data

,Animal,Speed,Name,Age,Experience,Company,Salary
0,Lion,100,John,27,1,Google,1200
1,tiger,180,peter,30,2,Meta,1400
2,Lion,90,steve,32,3,IBM,1000
3,tiger,190,brad,35,4,MSFT,1800
4,Lion,80,smith,29,5,Wipro,1100
5,tiger,200,Hussy,41,6,Cognizant,900
